csrc/optimization/consecutive_cast_pass.cpp

```diff
@@ -0,0 +1,151 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/utils.h>
+#include <optimization/consecutive_cast_pass.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+bool isSameDtypeCategory(const DataType& input_t, const DataType& output_t) {
+  if ((isIntegralType(input_t) && isIntegralType(output_t)) ||
+      (isFloatingPointType(input_t) && isFloatingPointType(output_t)) ||
+      (isComplexType(input_t) && isComplexType(output_t))) {
+    return true;
+  }
+  return false;
+}
+
+// note: returns
+//  - -1 : v0 contains strictly more information than v1;
+//  - 0  : a complex case, where each v0 and v1 isn't a super set of the other;
+//  - 1  : v0 and v1 has the same dtype;
+//  - 2  : v0 contains strictly less information than v1;
+int checkInformationLoss(Val* v0, Val* v1) {
+  auto dtype0 = v0->getDataType().value();
+  auto dtype1 = v1->getDataType().value();
+  if (dtype0 == dtype1) {
+    return 1;
+  }
+  if ((dtype0 == DataType::BFloat16 && dtype1 == DataType::Half) ||
+      (dtype1 == DataType::BFloat16 && dtype0 == DataType::Half)) {
+    return 0;
+  }
+  if (isWiderType(dtype0, dtype1)) {
+    return 2;
+  }
+  TORCH_INTERNAL_ASSERT(
+      isWiderType(dtype1, dtype0), "unrecognized cast category is encountered");
+  return -1;
+}
+
+void castOptimizationPass(Fusion* fusion) {
+  auto is_foldable_cast_op = [](Expr* expr) {
+    if (expr != nullptr && expr->isA<UnaryOp>()) {
+      auto op = expr->as<UnaryOp>();
+      if (op->getUnaryOpType() == UnaryOpType::Cast &&
+          isSameDtypeCategory(
+              expr->input(0)->getDataType().value(),
+              expr->output(0)->getDataType().value())) {
+        return true;
+      }
+    }
+    return false;
+  };
```

Comment:
Can you please move this as a standalone function with a comment defining what foldable means?